### PR TITLE
Added Streaming Sink for SQL DB

### PR DESCRIPTION
**Pull request by romitgirdhar**
Added Streaming sink for SQL DB. This should enable users to write data to SQL DB or SQL Server in streaming mode (Structured Streaming). Only SQL Auth supported at this time.

### CODE REVIEW

**Comment by allenwux** [comment]
#sign-off

**Review by allenwux — APPROVED** [review]

File: src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    )
+    val password = writeConfig.get[String](SqlDBConfig.Password).getOrElse(
+      throw new IllegalArgumentException("password not found in Config")
+    )
+    val port = writeConfig.get[String](SqlDBConfig.PortNumber).getOrElse("1433")
+    val ignoreColumnNames = Try(writeConfig.get[String]("ignoreColumnNames").map(_.toBoolean).getOrElse(false)).getOrElse(false)
+
+    //Using regular writes
+    Class.forName(DRIVER_NAME)
+    connection = DriverManager.getConnection(createJDBCUrl(url, Some(port)) + ";database=" + db, user, password)
```
Comment (arvindshmicrosoft) [inline src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala]: Please reuse the connection mechanism used in existing code paths, so that other connection options like AAD authentication etc. can be used. The current implementation seems to impose only user + password authentication which is limiting for some users who want to use AAD.

File: src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala [hunk 2]
```diff
@@ -0,0 +1,193 @@
+/*
```
Comment (arvindshmicrosoft) [inline src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala]: Please review the license headers in the source files to ensure consistency with the rest of the repo (ASF vs. MIT)

File: src/test/scala/com/microsoft/azure/sqldb/spark/utils/SQLServerTestUtils.scala [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    val table = writeConfig.get[String](SqlDBConfig.DBTable).getOrElse(null)
+    if(table.equals(null)) {
+      return false
+    }
+    val conn = getConnection(writeConfig, true)
+    if(conn.equals(null)){
+      return false
+    }
+
+    var sql = "TRUNCATE TABLE " + table + ";"
```
Comment (arvindshmicrosoft) [inline src/test/scala/com/microsoft/azure/sqldb/spark/utils/SQLServerTestUtils.scala]: This is perhaps the first case where we are executing dynamically constructed SQL statement. A general good practice here would be to handle spaces and special characters within the 'table' variable by surrounding that within []. In this case it is probably okay given this is test code but in general you should watch out for strange characters in SQL object identifiers when constructing SQL like this.

File: src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala (hunk 1 above)
Comment (arvindshmicrosoft) [inline src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala]: Please reuse the connection mechanism used in existing code paths, so that other connection options like AAD authentication etc. can be used. The current implementation seems to impose only user + password authentication which is limiting for some users who want to use AAD.

File: src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (arvindshmicrosoft) [inline src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala]: Please review the license headers in the source files to ensure consistency with the rest of the repo (ASF vs. MIT)

File: src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLSinkProvider.scala [hunk 5]
```diff
@@ -0,0 +1,45 @@
+/*
```
Comment (arvindshmicrosoft) [inline src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLSinkProvider.scala]: ASF vs. MIT, please review other artifacts as well.

File: src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+          values += "?,"
+          schemaDatatype += col.dataType
+          accessorsArr += getAccessor(col.dataType)
+        })
+      }
+
+      var sql = ""
+
+      if(ignoreColumnNames) {
+        sql = "INSERT INTO " + table + " VALUES (" + values.substring(0, values.length - 1) + ");"
```
Comment (arvindshmicrosoft) [inline src/main/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLWriter.scala]: What if 'table' has a space, or reserved characters in the object identifier? On a similar note, what if the table has a schema name (in SQL you can have tables in other schemas). Since we are constructing dynamic SQL these edge cases might need to be handled.

File: src/test/scala/com/microsoft/azure/sqldb/spark/utils/SQLServerTestUtils.scala [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    val port = writeConfig.get[String](SqlDBConfig.PortNumber).getOrElse("1433")
+    if(db.equals(null) || url.equals(null)){
+      return null
+    }
+    Class.forName(DRIVER_NAME)
+    var dbConnect = ""
+    if(connectToDB){
+      dbConnect = ";database="+db
+    }
+    var conn: Connection = DriverManager.getConnection(createJDBCUrl(url, Some(port))+ dbConnect, user, password)
```
Comment (arvindshmicrosoft) [inline src/test/scala/com/microsoft/azure/sqldb/spark/utils/SQLServerTestUtils.scala]: Similar thing to other comment, try reusing existing connection code so that we can support all the config cases like AAD authentication etc.

File: src/test/scala/com/microsoft/azure/sqldb/spark/utils/SQLServerTestUtils.scala [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  private def preFlightChecks(config: Map[String, String]): Boolean = {
+    var writeConfig = Config(config)
+    val conn = getConnection(writeConfig, false)
+    val dbName = writeConfig.get[String](SqlDBConfig.DatabaseName).get
+    if(conn.equals(null)){
+      return false
+    }
+
+    //Check if Server and Database Exist
+    var sql = "SELECT name from master.dbo.sysdatabases where name='"+ dbName + "' OR [name]='"+ dbName + "'"
```
Comment (arvindshmicrosoft) [inline src/test/scala/com/microsoft/azure/sqldb/spark/utils/SQLServerTestUtils.scala]: use sys.databases, no longer recommended to explicitly reference sysdatabases (it is an older view). and similar comments on handling database names - need to potentially handle embedded spaces / special chars in the dbName.

File: src/test/scala/com/microsoft/azure/sqldb/spark/utils/SQLServerTestUtils.scala [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    if(table.equals(null)) {
+      return false
+    }
+    val conn = getConnection(writeConfig, true)
+    if(conn.equals(null)){
+      return false
+    }
+
+    // var sql = "IF EXISTS(SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = '" + table + "') DROP TABLE " + table + ";"
+    var sql = "DROP TABLE IF EXISTS " + table   //Supported SQL Server 2016 and beyond
```
Comment (arvindshmicrosoft) [inline src/test/scala/com/microsoft/azure/sqldb/spark/utils/SQLServerTestUtils.scala]: given this in a test, I will let this fly. otherwise in general there are still quite a few customers who use older versions of SQL hence the first way (checking IF EXISTS) might be a better option, but that's fine. No need to change this, just FYI.

File: src/test/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLSinkTest.scala [hunk 10]
```diff
@@ -0,0 +1,457 @@
+package com.microsoft.azure.sqldb.spark.sql.streaming
```
Comment (arvindshmicrosoft) [inline src/test/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLSinkTest.scala]: pls. add appropriate license

File: src/test/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLSinkTest.scala [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+  implicit val sparkSession: SparkSession = SparkSession.builder().config(sc.getConf).getOrCreate() //enableHiveSupport().
+  import sparkSession.implicits._
+
+  override val streamingTimeout: Span = 60.seconds
+  val url = sys.env.getOrElse("DB_SERVER", "localhost")
+  val database = sys.env.getOrElse("DATABASE_NAME", "test1")
+  val user = sys.env.getOrElse("DB_USERNAME", "test")
+  val password = sys.env.getOrElse("DB_PASSWORD", "test")
+  val dbTable = "dbo.newtesttable"
+  val portNum = sys.env.getOrElse("PORT_NUM", "58502")
```
Comment (arvindshmicrosoft) [inline src/test/scala/com/microsoft/azure/sqldb/spark/sql/streaming/SQLSinkTest.scala]: Why this magic number 58502?

**Review by arvindshmicrosoft — CHANGES_REQUESTED** [review]
Thanks Romit for your work! A few comments inline.

**Comment by renepajta** [comment]
Hi all. Great job with Streaming sink. When can we see this merged to master? Thank you.

**Comment by arvindshmicrosoft** [comment]
@renepajta this work will not merge into this repo as it has not been actively maintained (other than a few minor updates) over the last year and half or more. If you'd like to see this with the [newer connector](https://github.com/microsoft/sql-spark-connector/) please feel free to open an issue on that project, so that the project owners can triage and consider.

**Comment by arvindshmicrosoft** [comment]
Closing due to inactivity and the inability to accept a large change like this on this non-maintained project. Closing this PR and cc @cchighman who was considering doing something similar for the [new connector](https://github.com/microsoft/sql-spark-connector).
